impl/src/main/kotlin/com/android/declarative/internal/variantApi/AndroidComponentsParser.kt

```diff
@@ -0,0 +1,108 @@
+
+/**
+ * Specialized parser responsible for parsing the 'androidComponents' elements
+ * in declarative file.
+ */
+class AndroidComponentsParser(
+    private val project: Project,
+    private val issueLogger: IssueLogger =
+        IssueLogger(false, LoggerWrapper(project.logger)),
+): DeclarativeFileParser {
+
+    override fun <T: Any> parse(table: TomlTable, type: KClass<out T>, extension: T) {
+        // the extension is a subclass of AndroidComponentsExtension, its type parameters will tell me
+        // type types of the VariantBuilder and Variant object I am dealing with.
+        val superType = type.supertypes.first { superType ->
+            superType.jvmErasure.simpleName == "AndroidComponentsExtension"
+        }
+
+        // now the superType arguments are as follows
+        // superType.arguments[0] is the DSL common extension type
+        // superType.arguments[1] is the VariantBuilder type
+        // superType.arguments[2] is the Variant type.
+        val variantBuilderType = superType.arguments[1].type
+            ?: throw RuntimeException("Cannot determine the VariantBuilder type for extension $type")
+
+        val variantType = superType.arguments[2].type
+            ?: throw RuntimeException("Cannot determine the Variant type for extension $type")
+
+        // It is not necessary to cast to the AndroidComponentsExtension subtype as I am only
+        // interested in invoking methods that are all defined in the AndroidComponentsExtension interface itself.
+        val typedExtension = AndroidComponentsExtension::class.java.cast(extension)
+
+        // order does not matter, it's just registering a callback at this point which will be called by AGP
+        // in the right order.
+        handleCallback(table, typedExtension, "beforeVariants", variantBuilderType)
+        { tomlDeclaration, variantApiType, selector ->
+            if (selector == null) {
+                typedExtension.beforeVariants { variantBuilder ->
+                    parse(tomlDeclaration, variantBuilder, variantApiType)
+                }
+            } else {
+                typedExtension.beforeVariants(selector) { variantBuilder ->
+                    parse(tomlDeclaration, variantBuilder, variantApiType)
+                }
+            }
+        }
+        handleCallback(table, typedExtension, "onVariants", variantType) {
+                tomlDeclaration, variantApiType, selector ->
+            if (selector == null) {
+                typedExtension.onVariants { variant ->
+                    parse(tomlDeclaration, variant, variantApiType)
+                }
+            } else {
+                typedExtension.onVariants(selector) { variant ->
+                    parse(tomlDeclaration, variant, variantApiType)
+                }
+            }
+        }
+    }
+
+    private fun handleCallback(
+        table: TomlTable,
+        typedExtension: AndroidComponentsExtension<*, *, *>,
+        variantApiName: String,
```

Comment:
maybe add a comment with examples of variantApiName; e.g., "onVariants", "beforeVariants"... it wasn't obvious to me from the name/code.

Reply:
Maybe even we can do enum as we have fixed choices anyway?

Reply:
Added comment
I'll think about enum once I add the other ones like `finalizeDsl`